DIPs/DIP1005.md

```diff
@@ -0,0 +1,781 @@
+
+|Aggregate|Imports (unittest)|Imports (compile)|Imports (top)|
+|---|---|---|---|
+|Median|30|10|8|
+|Average|28.0|13.8|10.5|
+
+Another matter we investigated is how reducing top-level imports influences build times and the size of the object files produced. We do not have an experimental implementation of this DIP, so measuring impact directly was not possible. We did the converse experiment---adding top-level imports.
+
+In a separate branch of the standard library code, for each module we added all nested imports back to the top level. Some hand-editing was needed after that because of clashes in symbol names. Also, some imports needed to be removed because of circular dependencies and related limitations in the language's design and implementation. The resulting setup can be seen in [PR4992](https://github.com/dlang/phobos/pull/4992). Then for each module in the standard library we compiled one file that consists of exactly one `import` declaration, monitoring compile time and object file size. Appendix B displays build times and size of object files produced by this experiment.
+
+|Aggregate|Time (top-level)|Object size (top-level)|Time (nested)|Object size (nested)|
+|---|---|---|---|---|
+|Median|320ms|13788|32ms|4296|
+|Average|287.6ms|13437.2|64.6ms|5734.1|
+
+As expected, the experiment shows that both build times and object file sizes were improved by moving imports away from the top level. We estimate that eliminating the 10.5x slack dependency fan-in will bring import costs down to negligible and also bring object file size down.
+
+The numbers show that good engineering improves matters significantly,  as expected. However, the resulting state of affairs is far from optimum. On average, for one standard library import, 10.5 more modules in the standard library are imported. If the import is not used, and assuming in first approximation that the total overhead of one import is proportional with the total number of modules transitively imported, we're looking at an order of magnitude overhead in dependency fan-in. As artifacts in the module do get used, dependencies are realized and the waste goes down. Further improvements would require a major project reorganization. We conclude that local imports are not enough to ensure an efficient dependency structure, even after discounting the other claimed advantages for maintainability, documentation, and code clarity.
+
+### Inline `import`s
+
+We propose an addition to the D language that allows the use of the keyword `import` as part of any function and aggregate declaration. When that syntax is used, it instructs the compiler to execute the import before looking up any names in the declaration. To clarify by means of example, the previous declarations would be rewritten as:
+
+```d
+with (import std.stdio) void process(File input) ;
+with (import std.range) struct Buffered(Range) if (isInputRange!Range)
```

Comment:
Is that always a private import? So another module using Buffered won't get the rest of std.range?
Would some `with (public import std.range)` make sense, e.g. to add `.front` and similar primitives.